lib/js/jslib.js

```diff
@@ -1,187 +1,255 @@
+        curL = curL._tail;
+        curR = curR._tail;
 
-function _lsReplicate (n, item) {
-  var out = lsNil;
-  while (n > 0) {
-    out = _lsCons(item, out);
-    --n;
-  }
-  return out;
-}
-const lsReplicate = _lsReplicate;
+      }
+      return LINKEDLIST.isNil(curL) && LINKEDLIST.isNil(curR)
+    },
 
-// faster than _lsReplicate? not really?
-function _lsRepeat (n, x) {
-  var arr = [ ];
-  var pattern = [ x ];
-  while (n > 0) {
-    if (n & 1) {
-      arr = arr.concat(pattern);
-    }
-    n >>= 1;
-    pattern = pattern.concat(pattern);
-  }
-  return _lsFromArray(arr);
-}
-const lsRepeat = _lsRepeat;
+    head: function(v) {
+      DEBUG.assert(TYPES.isLinkedList(v), "Not a linked list"); 
+      return LINKEDLIST.isNil(v) ? _error('head') : v._head; 
+      }, 
 
-function _ls (arr) {
-	var out = lsNil;
-	for (var i = arr.length - 1; i >= 0; --i) {
-		out = _lsCons(arr[i], out);
-	}
-	return out;
-}
-const ls = _ls;
+    tail: function(v) {
+      DEBUG.assert(TYPES.isLinkedList(v), "Not a linked list"); 
+      return LINKEDLIST.isNil(v) ? _error('tail') : v._tail; 
+      },
 
-function _lsMinimum(xs) {
-  var currentMin = _lsHead(xs);
-  while (xs !== null) {
-    if (xs._head < currentMin) currentMin = xs._head;
-    xs = xs._tail;
-  }
-  return currentMin;
-}
-const lsMinimum = _lsMinimum;
+    last: function(xs) {
+      if (LINKEDLIST.isNil(xs)) {
+        return _error('last');
+      }
+      var out = xs._head;
+      while (!LINKEDLIST.isNil(xs)) {
+        out = xs._head;
+        xs = xs._tail;
+      }
+      return out;
+    },
+
+      revAppend: function(xs, ys) {
+        var out = ys;
+        while (!LINKEDLIST.isNil(xs)) {
+          out = LINKEDLIST.Cons(head(xs), out);
+          xs = LINKEDLIST.tail(xs);
+        }
+        return out;
+      },
+
+    append: function(xs, ys) {
+      DEBUG.assert(TYPES.isLinkedList(xs) && TYPES.isLinkedList(ys), "Not linked lists"); 
+      if (LINKEDLIST.isNil(xs)) {
+        return ys;
+      }
+      const origxs = xs;
+      const rootEl = _Cons(xs._head, Nil);
+      let curr = rootEl;
+      xs = xs._tail;
+      while (!LINKEDLIST.isNil(xs)) {
+        curr._tail = _Cons(xs._head, Nil);
+        xs = xs._tail;
+        curr = curr._tail;
+      }
+      curr._tail = ys;
+      return rootEl;
+    },
+
+    at: function(xs, i) {
+      var out;
+      while (!LINKEDLIST.isNil(xs) && i >= 0) {
+        out = xs._head;
+        xs = xs._tail;
+        --i;
+      }
+      // FIXME should check i / index out of bounds here
+      return out;
+    },
+
+    empty: function(xs) {
+      return LINKEDLIST.isNil(xs);
+    },
+
+    zip: function(xs, ys) {
+      var arr = [ ];
+      while (!LINKEDLIST.isNil(xs) && !LINKEDLIST.isNil(ys)) {
+        arr.push({ "1": xs._head, "2": ys._head }); // { ctor:"_Tuple2", _0:x, _1:y }
+        xs = xs._tail;
+        ys = ys._tail;
+      }
+      return LINKEDLIST.lsFromArray(arr);
+    },
+
+    range:  function(a, b) {
```

Comment:
Same like a specialised version of `intRange` which we have special sugar for in the source syntax, i.e. `[1...42]`. I wonder whether `range` is used anywhere?

Reply:
I haven't found any occurences

Reply:
Interesting, maybe it is in fact dead code.